tc/c2/tc_op.h

```diff
@@ -95,38 +97,38 @@ class TcOp : public Operator<Context> {
   }
 
   virtual bool RunOnDevice() override {
-    // first, given the TC, define it in the executionEngine_
-    executionEngine_->define(tc_);
-
-    // now, given the input tensors, convert them to dlpack tensors so that
-    // we can call the compile command
-    std::vector<::tc::dlutils::DLTensorUPtr> inTensorUPtrs;
-    std::vector<const DLTensor*> inputDLTensors;
-    for (int idx = 0; idx < this->InputSize(); ++idx) {
-      auto dims = this->Input(idx).dims();
-      inTensorUPtrs.emplace_back(
-          dlpack::makeConstDLTensor(this->Input(idx), dims));
-      inputDLTensors.push_back(inTensorUPtrs.back().get());
+    if (!compiled_) {
```

Comment:
What if dimensions changed? You might need to store a cache of compiled kernels and compare dimensions. Or have a flag that tells TcOp that dimensions are fixed. Do you think we can do dimensions checks fast enough so we can actually do them all the time?

Reply:
This check is provided in ```executionEngine_->run``` which should be run under a ```checkSizes_``` flag instead of ```profile_``` flag. Without it you ```uncheckedRun``` and call the ```void*``` version.
I'd say it is too expensive to run all the time for low latency kernels but YMMV, measure it ?

@abadams was saying that we don't need 2 modes and that Halide + LLVM can compile a check function for you.
I think this would have been tricky to implement with the caching mechanism but in #307 we are killing it so there is an opportunity to stick that piece and drop ```uncheckedRun``` once and for all.

However, note that if your sizes don't match, we will have to raise an error for now: you must fully JIT TCs for now until support for more parametric TCs is implemented. Note that [this example](https://github.com/facebookresearch/TensorComprehensions/blob/master/tc/examples/tensordot.cc#L72) shows how to easily reuse pre-tuned options for different sizes, as-is, with reasonable perf, or as a starting point for a short tuning specialization run.

As @abadams had also suggested at some point we can also use that strategy for tuning for a neighborhood and synthesize parametric kernels. We should definitely explore that route if only to benchmark its feasibility in general and the performance differences we see in parametric vs non-parametric cases.

Reply:
#352